src/middleware.rs

```diff
@@ -17,13 +19,27 @@ impl VerifyApiKey {
 
 impl<S> Middleware<S> for VerifyApiKey {
     fn start(&self, req: &HttpRequest<S>) -> Result<Started> {
-        let secret = req.headers()
-            .get("x-beryl-secret")
-            .ok_or(ErrorUnauthorized(ParseError::Header))?
-            .to_str()
-            .map_err(ErrorUnauthorized)?;
+        let secret_header = req.headers().get(X_BERYL_SECRET);
+
+        let qp_secret_is_valid = {
+            let qry = req.query();
+            let secret_query_param = qry.get(X_BERYL_SECRET);
+
+            match secret_query_param {
```

Comment:
I think this block can be rewritten as
```rust
secret_query_param
  .map(|val| val == self.secret)
  .unwrap_or(false)
```
Let me know which way you prefer. I would generally use combinators instead of matches, but I'm fine with having match here. (possibly taking into account how the next block would be written using combinators).

Reply:
Just want to note that we might also be able to use `.or()` between the two `.get()`, which would short-circuit trying the url and the query. But I don't think it's necessary to use, it might not be clearer, and probably makes zero difference in terms of performance.

Reply:
Thanks @hwchen! I like the combinators approach (wasn't familiar previously with `unwrap_or`, very nice). Regarding the second point -- I'm not really concerned with the short circuiting and think that might be simpler from a readability standpoint as-is.